From 387230057d55fef80627f0ef30df65f968c1e1af Mon Sep 17 00:00:00 2001
From: Peter Swain <pswain@marvell.com>
Date: Thu, 16 May 2019 15:46:46 -0700
Subject: [PATCH 237/767] mmc: cavium: avoid single-slot startup issues

commit 01d87e93c8b7184082a23a8de6438d7711a64d2b from
git@git.assembla.com:cavium/WindRiver.linux.git

Previous code relied on particular state left by u-boot,
or a previous _probe attempt aborted with EPROBE_DEFER,
else it would program a clock-divisor of zero and hang MMC.

Do not cache emm_switch setting if it has clock divisor zero.
Do not use frequency params before they have been extracted
from devtree.
Do enforce lower frequency bound (400KHz) from mmc/core.

Change-Id: I2d19084d1ab57d0f98e2ce1ffcd6fb918228f53c
Signed-off-by: Peter Swain <pswain@marvell.com>
Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/8549
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/mmc/host/cavium.c | 7 +++++--
 drivers/mmc/host/cavium.h | 1 +
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index 17cf00930036..6c15aaf5ff6d 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -323,7 +323,7 @@ static void do_switch(struct cvm_mmc_host *host, u64 emm_switch)
 
 	check_switch_errors(host);
 
-	if (slot) {
+	if (slot && (emm_switch & MIO_EMM_SWITCH_CLK)) {
 		slot->cmd6_pending = false;
 		slot->cached_switch = emm_switch;
 	}
@@ -1329,6 +1329,8 @@ static void cvm_mmc_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 	clock = ios->clock;
 	max_f = max_supported_frequency(host);
 
+	if (clock < mmc->f_min)
+		clock = mmc->f_min;
 	if (clock > max_f)
 		clock = max_f;
 
@@ -1521,6 +1523,7 @@ static int cvm_mmc_init_lowlevel(struct cvm_mmc_slot *slot)
 	/* Make the changes take effect on this bus slot. */
 	set_bus_id(&emm_switch, slot->bus_id);
 	do_switch(host, emm_switch);
+	slot->cached_switch = emm_switch;
 	host->powered = true;
 
 	/*
@@ -1679,8 +1682,8 @@ int cvm_mmc_of_slot_probe(struct device *dev, struct cvm_mmc_host *host)
 
 	host->acquire_bus(host);
 	host->slot[id] = slot;
-	cvm_mmc_switch_to(slot);
 	cvm_mmc_init_lowlevel(slot);
+	cvm_mmc_switch_to(slot);
 	host->release_bus(host);
 
 	ret = mmc_add_host(mmc);
diff --git a/drivers/mmc/host/cavium.h b/drivers/mmc/host/cavium.h
index 60c42b1f6ee4..bfad450432c6 100644
--- a/drivers/mmc/host/cavium.h
+++ b/drivers/mmc/host/cavium.h
@@ -275,6 +275,7 @@ struct cvm_mmc_cr_mods {
 #define MIO_EMM_SWITCH_TIMING		GENMASK_ULL(50, 48)
 #define MIO_EMM_SWITCH_BUS_WIDTH	GENMASK_ULL(42, 40)
 #define MIO_EMM_SWITCH_POWER_CLASS	GENMASK_ULL(35, 32)
+#define MIO_EMM_SWITCH_CLK		GENMASK_ULL(31, 0)
 #define MIO_EMM_SWITCH_CLK_HI		GENMASK_ULL(31, 16)
 #define MIO_EMM_SWITCH_CLK_LO		GENMASK_ULL(15, 0)
 
-- 
2.31.1

